### DISCUSSION TITLE
Issues with pip install jax compatibility

### DISCUSSION THREAD

**Discussion by marcello-calabrese** [google/lightweight_mmm#14] [General]
Hi, i tried to install lightweight with pip install but after a while it seems there is an issue with jax and jax library compatibility.

And it gives me error and don't install the library.

Thanks for helping.

Marcello

**Comment by pabloduque0** [comment]
Hello Marcello! Thanks for reporting this.

Can you show what error are you getting? Did you have JAX installed prior to installing LMMM?

Best,

**Reply by marcello-calabrese** [reply]
thanks for the quick reply Pablo. No I didnt have JAX installed  because i created a brand new environment only for lightweight and then I pip installed. I will show you the error, i deleted the environment so i have to re create it again :-).  Thanks, Marcello

**Reply by marcello-calabrese** [reply]
Hi again, see below the error I get. I might be wrong but it seems that jax versions and compatibility. Thanks, Marcello


  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting flatbuffers~=1.12.0
  Using cached flatbuffers-1.12-py2.py3-none-any.whl (15 kB)
Collecting gast==0.3.3
  Using cached gast-0.3.3-py2.py3-none-any.whl (9.7 kB)
Collecting h5py~=2.10.0
  Using cached h5py-2.10.0-cp38-cp38-win_amd64.whl (2.5 MB)
Collecting typing-extensions~=3.7.4
  Using cached typing_extensions-3.7.4.3-py3-none-any.whl (22 kB)
Collecting termcolor~=1.1.0
  Using cached termcolor-1.1.0-py3-none-any.whl
Collecting absl-py
  Using cached absl_py-0.15.0-py3-none-any.whl (132 kB)
Collecting opt-einsum~=3.3.0
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
Requirement already satisfied: wheel~=0.35 in c:\users\marce\anaconda3\envs\gmmm\lib\site-packages (from tensorflow==2.4.1->lightweight_mmm) (0.37.1)
Collecting six~=1.15.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting tensorboard~=2.4
  Using cached tensorboard-2.9.1-py3-none-any.whl (5.8 MB)
Collecting wrapt~=1.12.1
  Using cached wrapt-1.12.1.tar.gz (27 kB)
Collecting tensorflow-estimator<2.5.0,>=2.4.0
  Using cached tensorflow_estimator-2.4.0-py2.py3-none-any.whl (462 kB)
Collecting astunparse~=1.6.3
  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting protobuf>=3.9.2
  Using cached protobuf-4.21.1-cp38-cp38-win_amd64.whl (524 kB)
Collecting grpcio~=1.32.0
  Using cached grpcio-1.32.0-cp38-cp38-win_amd64.whl (2.6 MB)
Collecting keras-preprocessing~=1.1.2
  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
Collecting numpy>=1.12
  Using cached numpy-1.19.5-cp38-cp38-win_amd64.whl (13.3 MB)
Collecting google-pasta~=0.2
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
Collecting numpyro>=0.8.0
  Using cached numpyro-0.9.1-py3-none-any.whl (283 kB)
  Using cached numpyro-0.9.0-py3-none-any.whl (282 kB)
  Using cached numpyro-0.8.0-py3-none-any.whl (264 kB)
INFO: pip is looking at multiple versions of numpy to determine which version is compatible with other requirements. This could take a while.
Collecting numpy>=1.12
  Using cached numpy-1.19.4-cp38-cp38-win_amd64.whl (13.0 MB)
  Using cached numpy-1.19.3-cp38-cp38-win_amd64.whl (13.3 MB)
  Using cached numpy-1.19.2-cp38-cp38-win_amd64.whl (13.0 MB)
INFO: pip is looking at multiple versions of kiwisolver to determine which version is compatible with other requirements. This could take a while.
Collecting kiwisolver>=1.0.1
  Using cached kiwisolver-1.4.1-cp38-cp38-win_amd64.whl (55 kB)
INFO: pip is looking at multiple versions of numpy to determine which version is compatible with other requirements. This could take a while.
  Using cached kiwisolver-1.4.0-cp38-cp38-win_amd64.whl (51 kB)
  Using cached kiwisolver-1.3.2-cp38-cp38-win_amd64.whl (52 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
  Using cached kiwisolver-1.3.1-cp38-cp38-win_amd64.whl (51 kB)
  Using cached kiwisolver-1.2.0-cp38-none-win_amd64.whl (58 kB)
  Using cached kiwisolver-1.1.0-cp38-none-win_amd64.whl (57 kB)
Requirement already satisfied: setuptools in c:\users\marce\anaconda3\envs\gmmm\lib\site-packages (from kiwisolver>=1.0.1->matplotlib==3.3.4->lightweight_mmm) (61.2.0)
  Using cached kiwisolver-1.0.1.tar.gz (31 kB)
INFO: pip is looking at multiple versions of kiwisolver to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of keras-preprocessing to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of jax to determine which version is compatible with other requirements. This could take a while.
Collecting jax>=0.2.21
  Using cached jax-0.3.12.tar.gz (947 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
  Using cached jax-0.3.11.tar.gz (947 kB)
  Using cached jax-0.3.10.tar.gz (939 kB)
  Using cached jax-0.3.9.tar.gz (937 kB)
  Using cached jax-0.3.8.tar.gz (935 kB)
  Using cached jax-0.3.7.tar.gz (944 kB)
  Using cached jax-0.3.6.tar.gz (936 kB)
INFO: pip is looking at multiple versions of keras-preprocessing to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of jax to determine which version is compatible with other requirements. This could take a while.
  Using cached jax-0.3.5.tar.gz (946 kB)
  Using cached jax-0.3.4.tar.gz (924 kB)
  Using cached jax-0.3.3.tar.gz (924 kB)
  Using cached jax-0.3.2.tar.gz (926 kB)
  Using cached jax-0.3.1-py3-none-any.whl
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
  Using cached jax-0.3.0.tar.gz (896 kB)
  Using cached jax-0.2.28.tar.gz (887 kB)
  Using cached jax-0.2.27.tar.gz (873 kB)
  Using cached jax-0.2.26.tar.gz (850 kB)
  Using cached jax-0.2.25.tar.gz (786 kB)
  Using cached jax-0.2.24.tar.gz (786 kB)
  Using cached jax-0.2.22.tar.gz (776 kB)
  Using cached jax-0.2.21.tar.gz (756 kB)
INFO: pip is looking at multiple versions of h5py to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of grpcio to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of google-pasta to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flatbuffers to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cycler to determine which version is compatible with other requirements. This could take a while.
Collecting cycler>=0.10
  Using cached cycler-0.10.0-py2.py3-none-any.whl (6.5 kB)
INFO: pip is looking at multiple versions of astunparse to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of absl-py to determine which version is compatible with other requirements. This could take a while.
Collecting absl-py
  Using cached absl_py-0.14.1-py3-none-any.whl (131 kB)
  Using cached absl_py-0.14.0-py3-none-any.whl (131 kB)
  Using cached absl_py-0.13.0-py3-none-any.whl (132 kB)
INFO: pip is looking at multiple versions of h5py to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of grpcio to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of google-pasta to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flatbuffers to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cycler to determine which version is compatible with other requirements. This could take a while.
  Using cached absl_py-0.12.0-py3-none-any.whl (129 kB)
  Using cached absl_py-0.11.0-py3-none-any.whl (127 kB)
  Using cached absl_py-0.10.0-py3-none-any.whl (127 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of gast to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of tensorflow to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of matplotlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of lightweight-mmm to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install lightweight-mmm==0.1.1, lightweight-mmm==0.1.2 and lightweight-mmm==0.1.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    lightweight-mmm 0.1.3 depends on jaxlib>=0.3.0
    lightweight-mmm 0.1.2 depends on jaxlib>=0.3.0
    lightweight-mmm 0.1.1 depends on jaxlib>=0.3.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies

**Reply by pabloduque0** [reply]
- What exact command did you run for installing?
- What is your PIP version?
- What OS are you running this? and what version?

**Comment by mahewashabdi** [comment]
Hi 
I am facing the same issue in the installation. Not able to install and getting the same conflict

thanks
Mahewash

**Reply by mahewashabdi** [reply]
I am using windows 11

**Reply by pabloduque0** [reply]
Hello @mahewashabdi !

JAX does not have very good compatibility with Windows, I would recommend you to try from a Linux machine or if you dont have access to one [Colaboratory](https://colab.sandbox.google.com/) might be a great option for you to try it :)

**Reply by marcello-calabrese** [reply]
ok thanks @pabloduque0 , so maybe this is the issue, I pip installed on my Windows 11. I will try to do it on my Linux/ubuntu virtual machine, I can do it on Colab but I prefer to have it in  my local machine. Thanks!!

**Reply by marcello-calabrese** [reply]
Folks!! It works on Ubuntu/Linux for me, i created a dedicated environment for it and pip installed lightweight
Hope it helps and thanks @pabloduque0

**Reply by mahewashabdi** [reply]
I implemented it on windows 11. you need to manually download the jaxlib whl file and then install it. It will be good to go.
